### ISSUE TITLE
`ttnn.where` (float32): a `-0.0` predicate selects the *true* branch — the SFPU kernel tests the raw bit pattern against zero, so `-0.0 != 0`, where torch and ttnn's own golden (`torch.ne(predicate, 0)`) select the false branch

### ISSUE THREAD

**Issue by beyondpremiere** [tenstorrent/tt-metal#56522]
Verified against tt-metal `main` @ `70f840f46adb33ec21ec8893f48ee06692648f75`, Blackhole ttsim v1.10.7, SFPI 7.76.0.
Evidence, probe and recorded rows: https://gist.github.com/beyondpremiere/172dc3264cb99ee638c3255e74289aa3

---

**Component / Area:** SFPU numerics — `tt_metal/tt-llk/tt_llk_{blackhole,wormhole_b0}/common/inc/sfpu/ckernel_sfpu_where.h`
**Issue Type:** Correctness

## Summary

`_calculate_where_` loads the predicate lane as a raw 32-bit integer (`mod0 = InstrModLoadStore::INT32`
for Float32, `ckernel_sfpu_where.h:29`) and decides the branch with

```
TTI_SFPSETCC(0, p_sfpu::LREG0, 0, sfpi::SFPSETCC_MOD1_LREG_EQ0);   // ckernel_sfpu_where.h:37 (and the
                                                                   //  SFPLOADMACRO variants at :54, :80)
```

That is an integer `== 0` on the bit pattern. `-0.0` is `0x80000000`, so it is classified as non-zero
and the kernel returns `true_value`. IEEE 754 says `-0.0 == 0`; `torch.where(cond.bool(), ...)` treats
`-0.0` as `False`; and ttnn's own reference does too — `ttnn/ttnn/operations/ternary.py:73,89`
normalises the predicate with `torch.ne(predicate, 0)` before calling `torch.where`. The op therefore
disagrees with its own golden on this input.

`ttnn.where` passes a float predicate straight to the kernel (`ternary.cpp:139-144` only typecasts the
dtype when it differs from the value tensors; there is no `!= 0` normalisation), so this reaches users
directly. Related to #43831 (same sign-bit-vs-zero pattern in the zero-comparison ops, fixed there) and
#56520 (`logical_not`, same residual).

## Observed

`sfpu_ternary_test.cpp`, `SfpuWhere`, Float32 in/out, `dest_acc=Yes`; predicate tile of 19 crafted
values, `true_value = 1000 + i`, `false_value = -(1000 + i)`, compared against
`torch.where(pred.bool(), a, b)`:

| predicate | bits | kernel picks | torch picks |
|---|---|---|---|
| `+0.0` | `0x00000000` | false | false |
| **`-0.0`** | `0x80000000` | **true** | false |
| `±1`, `±0.5`, `2`, `±1e-20`, `FLT_MAX` | — | true | true |
| `±inf` | — | true | true |
| NaN (`0x7fc00000`, `0xffc00000`, `0x7fc00001`) | — | true | true |
| `±min normal`, `±smallest subnormal` | — | true | true |

Only `-0.0` disagrees. `-0.0` is an ordinary product of float arithmetic (`0.0 * -x`, `-x` of a zero, a
negative product underflowing), so a predicate built from such a tensor silently flips those lanes.

## Proposed fix

For float layouts, drop the sign bit before the zero test. One instruction, no constant register:
shifting the pattern left by one turns `0x80000000` into `0` and leaves every non-zero magnitude
non-zero.

```cpp
// DISABLE_SFPLOADMACRO path, Float32 layout
TT_SFPLOAD(p_sfpu::LREG0, mod0, ADDR_MOD_7, offset0);
TT_SFPLOAD(p_sfpu::LREG1, mod0, ADDR_MOD_7, offset1);
TTI_SFPSHFT(1, p_sfpu::LREG0, p_sfpu::LREG0, 1);   // LREG0 <<= 1: -0.0 -> 0, everything else stays non-zero
TTI_SFPSETCC(0, p_sfpu::LREG0, 0, sfpi::SFPSETCC_MOD1_LREG_EQ0);
```

(replay length 6 → 7). **Validated on the simulator:** with that change, all 19 predicates above match
torch, `-0.0` included, and nothing else moves (`patch_where_signed_zero.py`). The `SFPLOADMACRO`
variants need the same shift inserted between the predicate load and the `SFPSETCC`. For the
`Float16_b` layout (`LO16`) the sign bit is bit 15, so the equivalent is a mask of `0x7FFF` rather
than a shift; I measured the Float32 path only. The Wormhole copy differs from Blackhole only in
`ADDR_MOD` constants and needs the same edit.

## Reproduction

`probe_where.py` (drop into `tt_metal/tt-llk/tests/python_tests/`, run with `--run-simulator --forked`,
`CHIP_ARCH=blackhole`). `baseline_where_specials.csv` and `fixed_where_specials.csv` are the recorded
outputs (predicate bits, kernel output, torch).

## Scope

Simulator-only (Blackhole ttsim). The mechanism is a bit-pattern compare in the kernel source, not a
simulator behaviour, but the result should be confirmed on hardware.

**Comment by shitobu** [comment]
Opened PR #56538 addressing this across Blackhole, Wormhole B0, and Quasar with regression tests in `test_where.py`.

- In `ckernel_sfpu_where.h` (Blackhole & Wormhole B0), Float32 shifts the predicate register left by 1 before `SFPSETCC_MOD1_LREG_EQ0` to flush `-0.0` (`0x80000000`) to `0` while keeping all non-zero float magnitudes non-zero.
- In Quasar `ckernel_sfpu_where.h`, evaluates `sfpi::abs(cond) != 0.0f`.
- Added unit test `test_ttnn_where_signed_zero_predicate` in `tests/ttnn/unit_tests/operations/eltwise/test_where.py` validating both `+0.0` and `-0.0` predicates against torch.
